src/utilities/createRouter.ts

```diff
@@ -90,21 +94,65 @@ export function createRouter<const T extends RouterRoutes>(routes: T, options: R
     }
   }
 
-  const push: RouterPush<T> = (source, options) => {
-    const url = resolve(source, options)
-
-    return update(url, { replace: options?.replace })
+  function push<TSource extends RouteKeysThatHaveRequireParams<T>>(
+    source: TSource,
+    params: RouteParamsByName<T, TSource>,
+    options?: RouterPushOptions
+  ): Promise<void>
+  function push<TSource extends RoutesKey<T>>(
+    source: TSource,
+    params?: RouteParamsByName<T, TSource>,
+    options?: RouterPushOptions
+  ): Promise<void>
+  function push(
+    source: Url,
+    options?: RouterPushOptions
+  ): Promise<void>
+  function push(
+    source: any,
+    paramsOrOptions?: any,
+    maybeOptions?: RouterPushOptions,
+  ): Promise<void> {
+    const options: RouterPushOptions = (isUrl(source) ? paramsOrOptions : maybeOptions) ?? {}
+    const url = resolve(source, paramsOrOptions, maybeOptions)
+
+    return update(url, { replace: options.replace })
   }
```

Comment:
I believe since you've defined the types with the overloads, if you use the RouterPush type here but add the implementation types same as you have here it'll work without having to redefine the overloads
```
  const push: RouterPush<T> = (source: any, paramsOrOptions?: any, maybeOptions?: RouterPushOptions): Promise<void> => {
    const options: RouterPushOptions = (isUrl(source) ? paramsOrOptions : maybeOptions) ?? {}
    const url = resolve(source, paramsOrOptions, maybeOptions)

    return update(url, { replace: options.replace })
  }
```

Reply:
ah yeah, good call. I was hoping to actually have the types inside of implementation but ended up needing to use any to get around type complaints